terraform/transform_attach_state.go

```diff
@@ -59,7 +59,9 @@ func (t *AttachStateTransformer) Transform(g *Graph) error {
 			continue
 		}
 
-		an.AttachResourceState(rs)
+		// make sure to attach a copy of the state, so instances can modify the
+		// same ResourceState.
+		an.AttachResourceState(rs.DeepCopy())
```

Comment:
I'm on board with making this a copy so that any later mutations are not visible in this object, but this makes me wonder... do we use this attached state anywhere except in other graph transforms? Is it possible that somewhere we're _relying_ on this being mutated during a walk for correct behavior? (I sure hope not, but this makes it seem like it was _possible_ for that to be true before.)

Reply:
I thought the same thing. No tests fail from this, but if we do encounter something relying on mutating the state, I think it will be easier to track that down than it will be to find a race.